File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Go/d40e3a87363b9200785a1ff37135694f.go
package main

import (
	"fmt"
	"math"
)

const mod = 998244353

type modint struct {
	x int
}

func newModInt(y int64) modint {
	y %= int64(mod)
	if y < 0 {
		y += int64(mod)
	}
	return modint{int(y)}
}

func (a modint) add(b modint) modint {
	c := a.x + b.x
	if c >= mod {
		c -= mod
	}
	return modint{c}
}

func (a modint) sub(b modint) modint {
	c := a.x - b.x
	if c < 0 {
		c += mod
	}
	return modint{c}
}

func (a modint) mul(b modint) modint {
	return modint{int(int64(a.x) * int64(b.x) % mod)}
}

func (a modint) pow(n int64) modint {
	ret := modint{1}
	mul := a
	for n > 0 {
		if n&1 == 1 {
			ret = ret.mul(mul)
		}
		mul = mul.mul(mul)
		n >>= 1
	}
	return ret
}

func (a modint) inv() modint {
	return a.pow(int64(mod - 2))
}

func (a modint) div(b modint) modint {
	return a.mul(b.inv())
}

func (a modint) equal(b modint) bool {
	return a.x == b.x
}

func (a modint) notEqual(b modint) bool {
	return a.x != b.x
}

func (a modint) lessThan(b modint) bool {
	return a.x < b.x
}

func (a modint) lessThanOrEqual(b modint) bool {
	return a.x <= b.x
}

func (a modint) greaterThan(b modint) bool {
	return a.x > b.x
}

func (a modint) greaterThanOrEqual(b modint) bool {
	return a.x >= b.x
}

func (a modint) String() string {
	return fmt.Sprintf("%d", a.x)
}

type fraction struct {
	num int64
	dom int64
}

func newFraction(n, d int64) fraction {
	if d == 0 {
		panic("division by zero")
	}
	if d < 0 {
		n = -n
		d = -d
	}
	g := gcd(abs(n), abs(d))
	n /= g
	d /= g
	return fraction{n, d}
}

func (a fraction) add(b fraction) fraction {
	num := a.num*b.dom + b.num*a.dom
	dom := a.dom * b.dom
	return newFraction(num, dom)
}

func (a fraction) sub(b fraction) fraction {
	num := a.num*b.dom - b.num*a.dom
	dom := a.dom * b.dom
	return newFraction(num, dom)
}

func (a fraction) mul(b fraction) fraction {
	num := a.num * b.num
	dom := a.dom * b.dom
	return newFraction(num, dom)
}

func (a fraction) div(b fraction) fraction {
	num := a.num * b.dom
	dom := a.dom * b.num
	return newFraction(num, dom)
}

func (a fraction) mulInt(k int64) fraction {
	num := a.num * k
	return newFraction(num, a.dom)
}

func (a fraction) divInt(k int64) fraction {
	dom := a.dom * k
	return newFraction(a.num, dom)
}

func (a fraction) lessThan(b fraction) bool {
	return a.num*b.dom < b.num*a.dom
}

func (a fraction) greaterThan(b fraction) bool {
	return a.num*b.dom > b.num*a.dom
}

func (a fraction) equal(b fraction) bool {
	return a.num*b.dom == b.num*a.dom
}

func (a fraction) notEqual(b fraction) bool {
	return a.num*b.dom != b.num*a.dom
}

func (a fraction) lessThanOrEqual(b fraction) bool {
	return a.num*b.dom <= b.num*a.dom
}

func (a fraction) greaterThanOrEqual(b fraction) bool {
	return a.num*b.dom >= b.num*a.dom
}

func abs(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

func gcd(a, b int64) int64 {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

type matrix struct {
	A [4][4]modint
}

func newMatrix(n int) matrix {
	var mat matrix
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if i == j {
				mat.A[i][j] = modint{1}
			} else {
				mat.A[i][j] = modint{0}
			}
		}
	}
	return mat
}

func (a matrix) add(b matrix) matrix {
	var c matrix
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			c.A[i][j] = a.A[i][j].add(b.A[i][j])
		}
	}
	return c
}

func (a matrix) sub(b matrix) matrix {
	var c matrix
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			c.A[i][j] = a.A[i][j].sub(b.A[i][j])
		}
	}
	return c
}

func (a matrix) mul(b matrix) matrix {
	var c matrix
	for i := 0; i < 4; i++ {
		for j := 0; j < 4; j++ {
			for k := 0; k < 4; k++ {
				c.A[i][j] = c.A[i][j].add(a.A[i][k].mul(b.A[k][j]))
			}
		}
	}
	return c
}

func (a matrix) pow(k int64) matrix {
	b := newMatrix(4)
	for i := 0; i < 4; i++ {
		b.A[i][i] = modint{1}
	}
	for k > 0 {
		if k&1 == 1 {
			b = b.mul(a)
		}
		a = a.mul(a)
		k >>= 1
	}
	return b
}

func (a matrix) String() string {
	var res string
	for i := 0; i < 4; i++ {
		res += "["
		for j := 0; j < 4; j++ {
			res += fmt.Sprintf("%v", a.A[i][j])
			if j != 3 {
				res += ","
			}
		}
		res += "]\n"
	}
	return res
}

type node struct {
	right int
	typ   int
}

func main() {
	var N int
	fmt.Scan(&N)
	X := make([]int, N)
	V := make([]int, N)
	P := make([]modint, N)
	Q := make([]modint, N)
	for i := 0; i < N; i++ {
		fmt.Scan(&X[i], &V[i], &P[i])
		P[i] = P[i].divInt(100)
		Q[i] = modint{1}.sub(P[i])
	}

	ev := make(map[fraction][]node)
	for i := 1; i < N; i++ {
		if V[i-1] < V[i] {
			ev[newFraction(int64(X[i-1]-X[i]), int64(V[i-1]-V[i]))] = append(ev[newFraction(int64(X[i-1]-X[i]), int64(V[i-1]-V[i]))], node{i, 0})
		}
		ev[newFraction(int64(X[i]-X[i-1]), int64(V[i-1]+V[i]))] = append(ev[newFraction(int64(X[i]-X[i-1]), int64(V[i-1]+V[i]))], node{i, 2})
		if V[i-1] > V[i] {
			ev[newFraction(int64(X[i-1]-X[i]), int64(V[i]-V[i-1]))] = append(ev[newFraction(int64(X[i-1]-X[i]), int64(V[i]-V[i-1]))], node{i, 1})
		}
	}

	type F func(a, b matrix) matrix
	f := func(a, b matrix) matrix {
		return a.mul(b)
	}

	seg := make([]matrix, 2*N)
	for i := 0; i < N; i++ {
		u := newMatrix(4)
		u.A[0][0] = Q[i]
		u.A[1][0] = Q[i]
		u.A[2][2] = Q[i]
		u.A[3][2] = Q[i]
		u.A[0][1] = P[i]
		u.A[1][1] = P[i]
		u.A[2][3] = P[i]
		u.A[3][3] = P[i]
		seg[i+N] = u
	}

	for i := N - 1; i > 0; i-- {
		seg[i] = f(seg[2*i], seg[2*i+1])
	}

	ret := modint{0}

	for p, vs := range ev {
		for _, tt := range vs {
			now := seg[tt.right]
			if tt.typ == 0 {
				now.A[0][0] = modint{0}
				now.A[0][2] = Q[tt.right]
			} else if tt.typ == 1 {
				now.A[1][1] = modint{0}
				now.A[1][3] = P[tt.right]
			} else {
				now.A[1][0] = modint{0}
				now.A[1][2] = Q[tt.right]
			}
			seg[tt.right] = now
		}
		uku := seg[1]
		sum := uku.A[0][2].add(uku.A[0][3])
		ret = ret.add(modint(p.num).divInt(p.dom).mul(sum))
		for _, tt := range vs {
			now := seg[tt.right]
			if tt.typ == 0 {
				now.A[0][2] = modint{0}
				now.A[2][2] = modint{0}
			} else if tt.typ == 1 {
				now.A[1][3] = modint{0}
				now.A[3][3] = modint{0}
			} else {
				now.A[1][2] = modint{0}
				now.A[3][2] = modint{0}
			}
			seg[tt.right] = now
		}
	}

	fmt.Println(ret)
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Java/d40e3a87363b9200785a1ff37135694f.java
import java.util.*;
import java.io.*;
import java.math.*;

class Main {
    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        public FastScanner() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

    static class IoSetup {
        IoSetup() {
            System.out.println("fixed");
            System.err.println("fixed");
        }
    }

    static class modint {
        int x;
        static final int mod = 998244353;

        modint() {
            x = 0;
        }

        modint(long y) {
            x = (int) (y >= 0 ? y % mod : (mod - (-y) % mod) % mod);
        }

        modint add(modint p) {
            if ((x += p.x) >= mod)
                x -= mod;
            return this;
        }

        modint subtract(modint p) {
            if ((x += mod - p.x) >= mod)
                x -= mod;
            return this;
        }

        modint multiply(modint p) {
            x = mul(x, p.x);
            return this;
        }

        modint divide(modint p) {
            multiply(p.inverse());
            return this;
        }

        modint negate() {
            return new modint(-x);
        }

        modint add(modint p) {
            return new modint(x + p.x);
        }

        modint subtract(modint p) {
            return new modint(x - p.x);
        }

        modint multiply(modint p) {
            return new modint(x * p.x);
        }

        modint divide(modint p) {
            return new modint(x * p.inverse().x);
        }

        boolean equals(modint p) {
            return x == p.x;
        }

        boolean notEquals(modint p) {
            return x != p.x;
        }

        boolean lessThan(modint p) {
            return x < p.x;
        }

        boolean greaterThan(modint p) {
            return x > p.x;
        }

        boolean lessThanOrEqual(modint p) {
            return x <= p.x;
        }

        boolean greaterThanOrEqual(modint p) {
            return x >= p.x;
        }

        modint inverse() {
            int a = x, b = mod, u = 1, v = 0, t;
            while (b > 0) {
                t = a / b;
                a -= t * b;
                u -= t * v;
                t = a;
                a = b;
                b = t;
                t = u;
                u = v;
                v = t;
            }
            return new modint(u);
        }

        modint pow(long n) {
            modint ret = new modint(1), mul = new modint(x);
            while (n > 0) {
                if ((n & 1) == 1)
                    ret.multiply(mul);
                mul.multiply(mul);
                n >>= 1;
            }
            return ret;
        }

        static int get_mod() {
            return mod;
        }

        static int mul(int a, int b) {
            long x = (long) a * b;
            long xh = (long) (x >> 32), xl = (long) x, d, m;
            asm("divl %4; \n\t" : "=a" (d), "=d" (m) : "d" (xh), "a" (xl), "r" (mod));
            return (int) m;
        }
    }

    static class fraction<T> {
        T num, dom;

        fraction() {
        }

        fraction(T n, T d) {
            num = n;
            dom = d;
            assert dom != 0;
            if (dom < 0) {
                num *= -1;
                dom *= -1;
            }
            T tmp = __gcd(abs(num), abs(dom));
            num /= tmp;
            dom /= tmp;
        }

        fraction add(fraction a) {
            return new fraction(num * a.dom + a.num * dom, dom * a.dom);
        }

        fraction subtract(fraction a) {
            return new fraction(num * a.dom - a.num * dom, dom * a.dom);
        }

        fraction multiply(fraction a) {
            return new fraction(num * a.num, dom * a.dom);
        }

        fraction divide(fraction a) {
            return new fraction(num * a.dom, dom * a.num);
        }

        fraction multiply(T k) {
            return new fraction(num * k, dom);
        }

        fraction divide(T k) {
            return new fraction(num, dom * k);
        }

        boolean lessThan(fraction a) {
            return num * a.dom < a.num * dom;
        }

        boolean greaterThan(fraction a) {
            return num * a.dom > a.num * dom;
        }

        boolean equals(fraction a) {
            return num * a.dom == a.num * dom;
        }

        boolean notEquals(fraction a) {
            return num * a.dom != a.num * dom;
        }

        boolean lessThanOrEqual(fraction a) {
            return num * a.dom <= a.num * dom;
        }

        boolean greaterThanOrEqual(fraction a) {
            return num * a.dom >= a.num * dom;
        }
    }

    static class Matrix<T> {
        ArrayList<ArrayList<T>> A;

        Matrix() {
        }

        Matrix(int n) {
            A = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                A.add(new ArrayList<>());
                for (int j = 0; j < n; j++) {
                    A.get(i).add(null);
                }
            }
        }

        int height() {
            return A.size();
        }

        int width() {
            return A.get(0).size();
        }

        T get(int i, int j) {
            return A.get(i).get(j);
        }

        void set(int i, int j, T val) {
            A.get(i).set(j, val);
        }

        static Matrix<Integer> I(int n) {
            Matrix<Integer> mat = new Matrix<>(n);
            for (int i = 0; i < n; i++) {
                mat.set(i, i, 1);
            }
            return mat;
        }

        Matrix<T> add(Matrix<T> B) {
            int n = height(), m = width();
            assert (n == B.height() && m == B.width());
            Matrix<T> C = new Matrix<>(n);
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    C.set(i, j, A.get(i).get(j) + B.get(i, j));
                }
            }
            return C;
        }

        Matrix<T> subtract(Matrix<T> B) {
            int n = height(), m = width();
            assert (n == B.height() && m == B.width());
            Matrix<T> C = new Matrix<>(n);
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    C.set(i, j, A.get(i).get(j) - B.get(i, j));
                }
            }
            return C;
        }

        Matrix<T> multiply(Matrix<T> B) {
            int n = height(), m = B.width(), p = width();
            assert (p == B.height());
            Matrix<T> C = new Matrix<>(n);
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    for (int k = 0; k < p; k++) {
                        C.set(i, j, C.get(i, j) + A.get(i).get(k) * B.get(k, j));
                    }
                }
            }
            return C;
        }

        Matrix<T> power(long k) {
            Matrix<T> B = Matrix.I(height());
            while (k > 0) {
                if ((k & 1) == 1)
                    B.multiply(this);
                multiply(this);
                k >>= 1;
            }
            return B;
        }

        T determinant() {
            Matrix<T> B = new Matrix<>(height());
            for (int i = 0; i < height(); i++) {
                for (int j = 0; j < width(); j++) {
                    B.set(i, j, A.get(i).get(j));
                }
            }
            assert (width() == height());
            T ret = 1;
            for (int i = 0; i < width(); i++) {
                int idx = -1;
                for (int j = i; j < width(); j++) {
                    if (B.get(j, i) != 0)
                        idx = j;
                }
                if (idx == -1)
                    return 0;
                if (i != idx) {
                    ret *= -1;
                    for (int j = 0; j < width(); j++) {
                        T tmp = B.get(i, j);
                        B.set(i, j, B.get(idx, j));
                        B.set(idx, j, tmp);
                    }
                }
                ret *= B.get(i, i);
                T vv = B.get(i, i);
                for (int j = 0; j < width(); j++) {
                    B.set(i, j, B.get(i, j) / vv);
                }
                for (int j = i + 1; j < width(); j++) {
                    T a = B.get(j, i);
                    for (int k = 0; k < width(); k++) {
                        B.set(j, k, B.get(j, k) - B.get(i, k) * a);
                    }
                }
            }
            return ret;
        }
    }

    static class Node {
        int right, type;

        Node(int right, int type) {
            this.right = right;
            this.type = type;
        }
    }

    static class SegmentTree {
        int sz;
        ArrayList<Matrix<modint>> seg;
        Matrix<modint> e;

        SegmentTree(int n, Matrix<modint> e) {
            sz = 1;
            while (sz < n)
                sz <<= 1;
            seg = new ArrayList<>(2 * sz);
            for (int i = 0; i < 2 * sz; i++) {
                seg.add(e);
            }
            this.e = e;
        }

        void set(int k, Matrix<modint> x) {
            seg.set(k + sz, x);
        }

        void build() {
            for (int k = sz - 1; k > 0; k--) {
                seg.set(k, seg.get(2 * k).multiply(seg.get(2 * k + 1)));
            }
        }

        void update(int k, Matrix<modint> x) {
            k += sz;
            seg.set(k, x);
            while ((k >>= 1) > 0) {
                seg.set(k, seg.get(2 * k).multiply(seg.get(2 * k + 1)));
            }
        }

        Matrix<modint> query(int a, int b) {
            Matrix<modint> L = e, R = e;
            for (a += sz, b += sz; a < b; a >>= 1, b >>= 1) {
                if ((a & 1) == 1)
                    L = L.multiply(seg.get(a++));
                if ((b & 1) == 1)
                    R = seg.get(--b).multiply(R);
            }
            return L.multiply(R);
        }
    }

    public static void main(String[] args) {
        FastScanner sc = new FastScanner();
        int N = sc.nextInt();
        int[] X = new int[N];
        int[] V = new int[N];
        modint[] P = new modint[N];
        modint[] Q = new modint[N];
        for (int i = 0; i < N; i++) {
            X[i] = sc.nextInt();
            V[i] = sc.nextInt();
            P[i] = new modint(sc.nextLong());
            P[i] = P[i].divide(100);
            Q[i] = new modint(1).subtract(P[i]);
        }

        TreeMap<fraction<Long>, ArrayList<Node>> ev = new TreeMap<>();
        for (int i = 1; i < N; i++) {
            if (V[i - 1] < V[i]) {
                ev.computeIfAbsent(new fraction<>(X[i - 1] - X[i], V[i - 1] - V[i]), k -> new ArrayList<>())
                        .add(new Node(i, 0));
            }

            ev.computeIfAbsent(new fraction<>(X[i] - X[i - 1], V[i - 1] + V[i]), k -> new ArrayList<>())
                    .add(new Node(i, 2));

            if (V[i - 1] > V[i]) {
                ev.computeIfAbsent(new fraction<>(X[i - 1] - X[i], V[i] - V[i - 1]), k -> new ArrayList<>())
                        .add(new Node(i, 1));
            }
        }

        Matrix<modint> e = Matrix.I(4);
        SegmentTree seg = new SegmentTree(N, e);
        for (int i = 0; i < N; i++) {
            Matrix<modint> u = new Matrix<>(e);
            u.set(0, 0, Q[i]);
            u.set(1, 0, Q[i]);
            u.set(2, 2, Q[i]);
            u.set(3, 2, Q[i]);
            u.set(0, 1, P[i]);
            u.set(1, 1, P[i]);
            u.set(2, 3, P[i]);
            u.set(3, 3, P[i]);
            seg.set(i, u);
        }
        seg.build();

        modint ret = new modint(0);

        for (Map.Entry<fraction<Long>, ArrayList<Node>> entry : ev.entrySet()) {
            ArrayList<Node> vs = entry.getValue();
            for (Node tt : vs) {
                Matrix<modint> now = seg.query(tt.right, tt.right);
                if (tt.type == 0) {
                    now.set(0, 0, new modint(0));
                    now.set(0, 2, Q[tt.right]);
                } else if (tt.type == 1) {
                    now.set(1, 1, new modint(0));
                    now.set(1, 3, P[tt.right]);
                } else {
                    now.set(1, 0, new modint(0));
                    now.set(1, 2, Q[tt.right]);
                }
                seg.update(tt.right, now);
            }
            Matrix<modint> uku = seg.query(0, 1);
            modint sum = uku.get(0, 2).add(uku.get(0, 3));
            ret.add(entry.getKey().num).divide(entry.getKey().dom).multiply(sum);
            for (Node tt : vs) {
                Matrix<modint> now = seg.query(tt.right, tt.right);
                if (tt.type == 0) {
                    now.set(0, 2, new modint(0));
                    now.set(2, 2, new modint(0));
                } else if (tt.type == 1) {
                    now.set(1, 3, new modint(0));
                    now.set(3, 3, new modint(0));
                } else {
                    now.set(1, 2, new modint(0));
                    now.set(3, 2, new modint(0));
                }
                seg.update(tt.right, now);
            }
        }
        System.out.println(ret);
    }
}
